Modules/_json.c

```diff
@@ -628,40 +634,37 @@ py_encode_basestring(PyObject* Py_UNUSED(self), PyObject *pystr)
 }
 
 static void
-scanner_dealloc(PyObject *self)
+scanner_dealloc(PyScannerObject *self)
 {
+    PyTypeObject *tp = Py_TYPE(self);
     /* bpo-31095: UnTrack is needed before calling any callbacks */
     PyObject_GC_UnTrack(self);
     scanner_clear(self);
-    Py_TYPE(self)->tp_free(self);
+    freefunc free_func = PyType_GetSlot(tp, Py_tp_free);
+    free_func(self);
+    Py_DECREF(tp);
 }
 
 static int
-scanner_traverse(PyObject *self, visitproc visit, void *arg)
+scanner_traverse(PyScannerObject *self, visitproc visit, void *arg)
 {
-    PyScannerObject *s;
-    assert(PyScanner_Check(self));
```

Comment:
I would prefer to keep the assertion for safety. Same comment for other removed assertions.

See: https://en.wikipedia.org/wiki/Wikipedia:Chesterton's_fence

Don't remove an assertion if you don't know why it was added ;-)

Reply:
Oh. I missed the fact that PyScanner_Check() requires PyScannerType, but it's stored in the module state and there is no way to access the module state here :-( There would be two way to access the module state:

* Get it from Py_TYPE(self): sadly, currently, types don't store a reference to their module
* Get it with PyState_FindModule(). Sadly, this API doesn't work with modules using the multiphase initialization.

This problem is solved by the PEP 573 which has just been accepted, but not implemented yet.

I don't think that it's worth it to keep the code commented until the PEP 573 is implemented. These assertions were just for sanity checks, but it's unlikely that we get an object of a different type here. I mean, that it's not worth it to bother with these assertions.